dhis-2/dhis-services/dhis-service-expression-parser/src/main/java/org/hisp/dhis/expressionparser/ExpressionItemsVisitor.java

```diff
@@ -0,0 +1,297 @@
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import org.antlr.v4.runtime.tree.ParseTree;
+import org.hisp.dhis.common.DimensionItemType;
+import org.hisp.dhis.common.DimensionService;
+import org.hisp.dhis.common.DimensionalItemObject;
+import org.hisp.dhis.constant.ConstantService;
+import org.hisp.dhis.organisationunit.OrganisationUnitGroup;
+import org.hisp.dhis.organisationunit.OrganisationUnitGroupService;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import static org.hisp.dhis.common.DimensionItemType.*;
+import static org.hisp.dhis.expression.ExpressionService.DAYS_DESCRIPTION;
+import static org.hisp.dhis.expression.ExpressionService.SYMBOL_DAYS;
+import static org.hisp.dhis.expressionparser.generated.ExpressionParser.*;
+
+/**
+ * ANTLR parse tree visitor to find the expression items in an expression,
+ * find the oranisation unit groups in an expression, and also used to construct
+ * an expression description by filling in the names of the items.
+ * <p/>
+ * Uses the ANTLR visitor partern.
+ *
+ * @author Jim Grace
+ */
+public class ExpressionItemsVisitor
+    extends ExpressionVisitor
+{
+    @Autowired
+    private DimensionService dimensionService;
+
+    @Autowired
+    private OrganisationUnitGroupService organisationUnitGroupService;
+
+    private Set<DimensionalItemObject> dimensionalItemObjects = null;
+
+    private Set<OrganisationUnitGroup> orgUnitGroupsNeeded = null;
+
+    public Set<DimensionalItemObject> getDimensionalItemObjects( ParseTree parseTree,
+        DimensionService _dimensionService )
+    {
+        //TODO: Why doesn't the @Autowired value work?
+        dimensionService = _dimensionService;
```

Comment:
Service must be pass in to the constructor

Reply:
Done.